packages/loopring_v3/contracts/amm/libamm/AmmExitRequest.sol

```diff
@@ -0,0 +1,174 @@
+
+    bytes32 constant public WITHDRAW_TYPEHASH = keccak256(
+        "Withdraw(address owner,uint256[] amounts,uint256 validUntil,uint256 nonce)"
+    );
+
+    function unlock(AmmData.State storage S)
+        internal
+        returns (uint lockedUntil)
+    {
+        require(S.lockedUntil[msg.sender] == 0, "UNLOCKED_ALREADY");
+
+        lockedUntil = block.timestamp + AmmData.MIN_TIME_TO_UNLOCK();
+        S.lockedUntil[msg.sender] = lockedUntil;
+    }
+
+    function withdrawFromPool(
+        AmmData.State storage S,
+        uint[]       calldata amounts,
+        bytes        calldata signature, // signature from Exchange operator
+        uint                  validUntil
+        )
+        public
+        returns (uint[] memory amountOuts)
+    {
+        uint size = S.tokens.length;
+        require(amounts.length == size + 1, "INVALID_DATA");
+
+        _proxcessExchangeWithdrawalApprovedWithdrawals(S);
+
+        bool approvedByOperator = _checkOperatorApproval(
+            S, amounts, signature, validUntil
+        );
+
+        amountOuts = new uint[](size + 1);
+        amountOuts[0] = _withdrawToken(
+            S, address(this), amounts[0], approvedByOperator
+        );
+
+        for (uint i = 0; i < size; i++) {
+            amountOuts[i + 1] = _withdrawToken(
+                S, S.tokens[i].addr, amounts[i + 1], approvedByOperator
+            );
+        }
+    }
+
+    function exitPool(
+        AmmData.State storage S,
+        uint                  poolAmountIn,
+        uint96[]     calldata minAmountsOut,
+        bool                  toLayer2
+        )
+        public
+        returns(AmmData.PoolExit memory exit)
+    {
+        require(minAmountsOut.length == S.tokens.length, "INVALID_DATA");
+
+        // To make the the available liqudity tokens cannot suddenly change
+        // we keep track of when onchain exits (which need to be processed) are pending.
+        require(S.isExiting[msg.sender] == false, "ALREADY_EXITING");
+        S.isExiting[msg.sender] = true;
+
+        exit = AmmData.PoolExit({
+            owner: msg.sender,
+            toLayer2: toLayer2,
+            poolAmountIn: poolAmountIn,
+            minAmountsOut: minAmountsOut,
+            storageIDs: new uint32[](0),
+            validUntil: 0xffffffff
+        });
+
+        // Approve the exit
+        bytes32 txHash = AmmUtil.hashPoolExit(S.domainSeperator, exit);
+        S.approvedTx[txHash] = block.timestamp;
+    }
+
+    function _checkOperatorApproval(
+        AmmData.State storage S,
+        uint[]       calldata amounts,
+        bytes        calldata signature, // signature from Exchange operator
+        uint                  validUntil
+        )
+        private
+        returns (bool)
+    {
+
+        // Check if we can withdraw without unlocking with an approval
+        // from the operator.
+        if (signature.length == 0) {
+            require(validUntil == 0, "INVALID_VALUE");
```

Comment:
Does this matter?

Reply:
Not really, just want to make sure there are no random data.